TYP: np.argmin and np.argmax overload changes

lvllvl opened the pull request:
Attempts to close #28641 
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

lvllvl committed:
TYP: np.argmin changes

jorenham commented on numpy/typing/tests/gh28641_argmin_argmax.py:
The NumPy type-tests are located at [numpy/typing/tests/data/reveal](https://github.com/numpy/numpy/tree/main/numpy/typing/tests/data/reveal) (acceptance- / true negatives) and [numpy/typing/tests/data/fail](https://github.com/numpy/numpy/tree/main/numpy/typing/tests/data/fail) (rejection / true positives)

lvllvl replied:
ok I will add tests in the right location.

I am going to reinsert the ones I removed. I removed them initially out of ignorance. But I wanted to remove this one: 

```python
 @overload
 def argmin(
     a: ArrayLike,
     axis: SupportsIndex | None = ...,
     out: None = ...,
     *,
     keepdims: bool = ...,
 ) -> Any: ...
```

Wouldn't a type `Any` potentially return a `float64`? And since we only want to return integers, then shouldn't this be removed?

Thanks for the review, I'll keep working on this.

jorenham replied:
No that one should stay. It handles the case where it cannot be determined whether it should return a scalar or an array.

jorenham commented on numpy/_core/fromnumeric.pyi:
```diff
@@ -111,6 +111,8 @@ _NumberOrObjectT = TypeVar("_NumberOrObjectT", bound=np.number | np.object_)
 _ArrayT = TypeVar("_ArrayT", bound=np.ndarray[Any, Any])
 _ShapeT = TypeVar("_ShapeT", bound=tuple[int, ...])
 _ShapeT_co = TypeVar("_ShapeT_co", bound=tuple[int, ...], covariant=True)
+_IndexArray = NDArray[np.signedinteger] | NDArray[np.unsignedinteger] | NDArray[np.bool_] # type alias for argmin / argmax
+_OutT = TypeVar("_OutT", bound=_IndexArray) # Type variable, must be assignable to _IndexArray
```
Since `_IndexArray` is only used once, it would be a bit cleaner to inline it. It's also fine if you want to keep it, but in that case it should then be annotated as `TypeAlias`.

Also, `_IndexArray` is currently too restrictive, because it would currently reject valid types like like `NDArray[bool_ | int_]`, `NDArray[int8 | uint8]`, and `NDArray[integer]`.

Additionally, it could help to have the name of the type parameter reflect the restriction on its upper bound. The name shows up in case of type-checker errors, e.g. when someone passes `NDArray[np.float64]`. And if the user then sees `_OutT`, it won't be very informative.

So given that, I'd personally probably go for something like:

```suggestion
_BoolOrIntArrayT = TypeVar("_BoolOrIntArrayT", bound=NDArray[np.integer | np.bool])
```

But there are probably many other good options too 🤷🏻

jorenham reviewed (changes requested):
Could you explain why you chose to remove some of the overloads, and why you modified the parameter types and return types of the other ones?

jorenham commented:
In case it wasn't clear yet: Fixing #28641 requires narrowing the `out` typevar bound of `np.arg{min,max}` (and `np.ndarray.arg{min,max}`) from `bound=NDArray[Any]` to `bound=NDArray[np.bool | np.integer]`. So there's no need to touch the overloads :)

jorenham commented on numpy/typing/tests/data/reveal/fromnumeric.pyi:
```diff
@@ -97,6 +97,11 @@ assert_type(np.argmin(AR_b, axis=0), Any)
 assert_type(np.argmin(AR_f4, out=AR_subclass), NDArraySubclass)
 
+OUT_I = np.empty((), dtype=np.intp)
+OUT_F = np.empty((), dtype=np.float64)
```
this seems redundant
```suggestion
```

jorenham commented on numpy/typing/tests/data/reveal/fromnumeric.pyi:
```diff
@@ -97,6 +97,11 @@ assert_type(np.argmin(AR_b, axis=0), Any)
 assert_type(np.argmin(AR_f4, axis=0), Any)
 assert_type(np.argmin(AR_f4, out=AR_subclass), NDArraySubclass)
 
+OUT_I = np.empty((), dtype=np.intp)
```
There's some shape-typing involved here, which is causing the mypy tests to fail. It's probably easiest to just use `AR_i8` instead.

lvllvl commented:
> In case it wasn't clear yet: Fixing #28641 requires narrowing the `out` typevar bound of `np.arg{min,max}` (and `np.ndarray.arg{min,max}`) from `bound=NDArray[Any]` to `bound=NDArray[np.bool | np.integer]`. So there's no need to touch the overloads :)

Thanks for your help, I adjusted tests that started to fail after I changed the stubs. 

Do you think additional testing is needed?

jorenham commented:
> Do you think additional testing is needed?

It wouldn't hurt to check that this now indeed rejects non-integer (or bool) arrays for `out` in `typing/tests/data/fail`

jorenham reviewed (commented):
The `np.arg{min,max}` functions look good now 👌🏻.

Since they're a a bit of a package deal with the `np.ndarray.arg{min,max}` methods, it would be best to also change those in the same way.

jorenham commented on numpy/__init__.pyi:
```diff
@@ -826,6 +826,7 @@ _CharacterItemT_co = TypeVar("_CharacterItemT_co", bound=_CharLike_co, default=_
 _TD64ItemT_co = TypeVar("_TD64ItemT_co", bound=dt.timedelta | int | None, default=dt.timedelta | int | None, covariant=True)
 _DT64ItemT_co = TypeVar("_DT64ItemT_co", bound=dt.date | int | None, default=dt.date | int | None, covariant=True)
 _TD64UnitT = TypeVar("_TD64UnitT", bound=_TD64Unit, default=_TD64Unit)
+_BoolOrIntArrayT = TypeVar("_BoolOrIntArrayT", bound=NDArray[np.integer | np.bool])
```
quick nitpick:

```suggestion
_BoolOrIntArrayT = TypeVar("_BoolOrIntArrayT", bound=NDArray[integer | np.bool])
```

lvllvl commented:
> I'd prefer it if you would run the test locally: CI isn't free, and each push notifies me :)

oh my mistake, ok will do!

jorenham commented:
thanks @lvllvl :)

lvllvl commented:
> thanks @lvllvl :)

Thanks for your help, I learned a lot! V fun.
